src/sdc11073/mdib/statecontainers.py

```diff
@@ -50,380 +98,496 @@ def update_from_other_container(self, other, skipped_properties=None):
 
-    def __repr__(self):
-        return f'{self.__class__.__name__} descriptorHandle="{self.DescriptorHandle}" StateVersion={self.StateVersion}'
+    def __repr__(self) -> str:
+        return f'{self.__class__.__name__} DescriptorHandle="{self.DescriptorHandle}" StateVersion={self.StateVersion}'
 
     @classmethod
-    def from_node(cls, node, descriptor_container=None):
+    def from_node(cls, node: Element,
+                  descriptor_container: AbstractDescriptorProtocol | None = None) -> AbstractStateContainer:
+        """Create an instance from XML node."""
         obj = cls(descriptor_container)
         obj.update_from_node(node)
         return obj
 
 
 class AbstractOperationStateContainer(AbstractStateContainer):
+    """Represents AbstractOperationState in BICEPS."""
+
     NODETYPE = pm.AbstractOperationState
     is_operational_state = True
-    OperatingMode = cp.EnumAttributeProperty('OperatingMode', default_py_value=pm_types.OperatingMode.ENABLED,
-                                             enum_cls=pm_types.OperatingMode)
+    OperatingMode: pm_types.OperatingMode = cp.EnumAttributeProperty('OperatingMode',
+                                                                     default_py_value=pm_types.OperatingMode.ENABLED,
+                                                                     enum_cls=pm_types.OperatingMode)
     _props = ('OperatingMode',)
 
 
 class SetValueOperationStateContainer(AbstractOperationStateContainer):
+    """Represents SetValueOperationState in BICEPS."""
+
     NODETYPE = pm.SetValueOperationState
-    AllowedRange = cp.SubElementListProperty(pm.AllowedRange, value_class=pm_types.Range)
+    AllowedRange: pm_types.Range = cp.SubElementListProperty(pm.AllowedRange, value_class=pm_types.Range)
     _props = ('AllowedRange',)
 
 
-class T_AllowedValues(pm_types.PropertyBasedPMType):  # pylint: disable=invalid-name
-    Value = cp.SubElementStringListProperty(pm.Value)
+class AllowedValuesType(pm_types.PropertyBasedPMType):  # pylint: disable=invalid-name
+    """Represents a list of values, in xml it is a list of pm.Value elements with one value as text."""
+
+    Value: list[str] = cp.SubElementStringListProperty(pm.Value)
     _props = ['Value']
 
-    def is_empty(self):
+    def is_empty(self) -> bool:
+        """Return True if Value is empty."""
         return self.Value is None or len(self.Value) == 0
 
 
 class SetStringOperationStateContainer(AbstractOperationStateContainer):
+    """Represents SetStringOperationState in BICEPS."""
+
     NODETYPE = pm.SetStringOperationState
-    AllowedValues = cp.SubElementWithSubElementListProperty(pm.AllowedValues,
-                                                            default_py_value=T_AllowedValues(),
-                                                            value_class=T_AllowedValues)
+    AllowedValues: AllowedValuesType = cp.SubElementWithSubElementListProperty(pm.AllowedValues,
```

Comment:
should this be a list where `AllowedValuesType` has one `Value`?

Reply:
No, AllowedValues has the list. The point is that the BICEPS schema requires at least one Value element if AllowedValues exists. This means the AllowedValue node must be omitted if there is no value. The SubElementWithSubElementListProperty implements that behavior.